### PR TITLE
Add gettext() package and encoding argument

### PR DESCRIPTION
**Pull request by cvwillegen**
This PR adds the possibility to translate strings in packages using gettext("message", "package", "encoding").

Implements #11637

### CODE REVIEW

File: runtime/doc/builtin.txt [hunk 1]
```diff
@@ -266,7 +268,8 @@ gettabvar({nr}, {varname} [, {def}])
 gettabwinvar({tabnr}, {winnr}, {name} [, {def}])
 				any	{name} in {winnr} in tab page {tabnr}
 gettagstack([{nr}])		Dict	get the tag stack of window {nr}
-gettext({text})			String	lookup translation of {text}
+gettext({text}[, {package}[, {encoding}]])
```
Comment (zeertzjq) [inline runtime/doc/builtin.txt]:
```suggestion
gettext({text} [, {package} [, {encoding}]])
```

File: runtime/doc/builtin.txt [hunk 2]
```diff
@@ -67,6 +67,8 @@ autocmd_get([{opts}])		List	return a list of autocmds
 balloon_gettext()		String	current text in the balloon
 balloon_show({expr})		none	show {expr} inside the balloon
 balloon_split({msg})		List	split {msg} as used for a balloon
+bindtextdomain({package}[, {path}])
```
Comment (zeertzjq) [inline runtime/doc/builtin.txt]:
```suggestion
bindtextdomain({package} [, {path}])
```

File: runtime/doc/builtin.txt [hunk 3]
```diff
@@ -1156,6 +1159,13 @@ balloon_split({msg})					*balloon_split()*
 <		{only available when compiled with the |+balloon_eval_term|
 		feature}
 
+bindtextdomain({package}[, {path}])			*bindtextdomain()*
```
Comment (zeertzjq) [inline runtime/doc/builtin.txt]:
```suggestion
bindtextdomain({package} [, {path}])			*bindtextdomain()*
```

File: runtime/doc/builtin.txt [hunk 4]
```diff
@@ -4276,7 +4286,7 @@ gettagstack([{winnr}])					*gettagstack()*
 			GetWinnr()->gettagstack()
 
 
-gettext({text})						*gettext()*
+gettext({text}[, {package}[, {encoding}]])		*gettext()*
```
Comment (zeertzjq) [inline runtime/doc/builtin.txt]:
```suggestion
gettext({text} [, {package} [, {encoding}]])		*gettext()*
```

**Comment by k-takata** [comment]
bindtextdomain() and bind_textdomain_codeset() change the setting of the current process.
So, they affect the translation of Vim itself.

Maybe we should change them temporarily only when calling gettext() from Vim script?

**Comment by cvwillegen** [comment]
> bindtextdomain() and bind_textdomain_codeset() change the setting of the current process. So, they affect the translation of Vim itself.
> 
> Maybe we should change them temporarily only when calling gettext() from Vim script?

bindtextdomain _should_ only be used with a specifier indicating the package (so not "vim". Perhaps we should check this...). The codeset is only changed during one call of the _() function, so that doesn't influence the translation of Vim itself.

**Comment by brammool** [comment]
> This PR adds the possibility to translate strings in packages using
> gettext("message", "package", "encoding").
> 
> Implements #11637 

Thanks for working on this.  For a new feature like this it is important
to "get it right".  Once plugin authors start using it we can't change
how it works without causing trouble.

It would be good to add a section to the help explaining, step-by-step,
how to add translations to a plugin.  ":help package-create" is a good
place for this.  This should make clear how bindtextdomain() needs to be
used and what package name to use there.  When this is written down we
may spot something that is complicated.  Making the
"src/po/tojavascript.vim" script available comes to mind.

Do we really need the {encoding} argument for gettext()?  Since strings
inside Vim script are always using 'encoding' I don't see how getting a
string in another encoding can be useful.

Instead of using textdomain() calls before and after using gettext(),
can we use dgettext()?  I'm not sure how widespread dgettext() is.  If
we need to use #ifdefs in case it is not available, that isn't more
elegant than using textdomain() twice.  However, I wonder what this
means for efficiency.  Does gettext() cache some information that is
lost when using another domain name?

For the f_gettext() implementation, if the second argument is NULL or an
empty string, it should behave like the argument is not present.  That
resembles what dgettext() does.

-- 
hundred-and-one symptoms of being an internet addict:
84. Books in your bookcase bear the names Bongo, WinSock and Inside OLE

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by RestorerZ** [comment]
@cvwillegen Thank you very much! «Сбылась мечта идиота!» :-)

**Comment by RestorerZ** [comment]
> Do we really need the {encoding} argument for gettext()? Since strings inside Vim script are always using 'encoding' I don't see how getting a string in another encoding can be useful.

Yes, Bram, I think it's necessary.
For example, this issue #11625 comes to mind. The problem there was never completely solved.
And with this argument, it would be possible, for example, to do something like this:
```
let enc = &enc
if "cp1251" == enc
   g:menutrans_set_lang_to = gettext("Включить для языка", "menu", "cp1251")
elseif "utf-8" == enc
   g:menutrans_set_lang_to = gettext("Включить для языка", "menu", "utf8")
endif

```
It remains only to understand whether this will work dynamically? As you correctly noted.

In general, the implementation of this function in such an expanded form would allow, for example, to take out of the package vim.mo everything related to optwin.vim and defaults.vim in separate packages for them. This would improve, in my opinion, the purity of the code ‐ each module has a separate package. Which, in turn, would allow, for example, to have translated messages and an interface, but the settings window is in English.

**Comment by brammool** [comment]
> > Do we really need the {encoding} argument for gettext()? Since
> > strings inside Vim script are always using 'encoding' I don't see
> > how getting a string in another encoding can be useful.
> 
> Yes, Bram, I think it's necessary.
> For example, this issue #11625 comes to mind. The problem there was
> never completely solved.
> And with this argument, it would be possible, for example, to do
> something like this:
> ```
> let enc = &enc
> if "cp1251" == enc
>    g:menutrans_set_lang_to = gettext("Включить для языка", "menu", "cp1251")
> elseif "utf-8" == enc
>    g:menutrans_set_lang_to = gettext("Включить для языка", "menu", "utf8")
> endif
> 
> ```

Using the value of 'encoding' for the encoding of the resulting text
would be the default.  Thus it would already do what you illustrated
without passing an extra argument with the encoding name.  The question
is if it would ever be useful to use another encoding than what is the
value of the 'encoding' option.

> It remains only to understand whether this will work dynamically? As
> you correctly noted.
> 
> In general, the implementation of this function in such an expanded
> form would allow, for example, to take out of the package vim.mo
> everything related to optwin.vim and defaults.vim in separate packages
> for them. This would improve, in my opinion, the purity of the code ‐
> each module has a separate package. Which, in turn, would allow, for
> example, to have translated messages and an interface, but the
> settings window is in English.

Having multiple packages makes the installation more complex.  I do not
see a real advantage in having multiple packages.  I have not heard
anyone ask for not using translations for the settings window only.
This looks like a theoretical thing, not a practical improvement that
users will want to use.

-- 
hundred-and-one symptoms of being an internet addict:
85. Choice between paying Compuserve bill and paying for kids education
    is a no brainer -- although a bit painful for your kids.

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by chrisbra** [comment]
is this ready?

File: Filelist [hunk 5]
```diff
@@ -219,6 +219,7 @@ SRC_ALL =	\
 		src/testdir/silent.wav \
 		src/testdir/popupbounce.vim \
 		src/testdir/crash/* \
+		src/testdir/ru_RU/LC_MESSAGES/__PACKAGE__.mo \
```
Comment (chrisbra) [inline Filelist:224]: I think we also need the __PACKAGE__.po file

File: runtime/doc/repeat.txt [hunk 6]
```diff
@@ -762,6 +765,35 @@ the command after changing the plugin help: >
 	:helptags path/start/foobar/doc
 	:helptags path/opt/fooextra/doc
 
+The messages that are in the lang/<lang_id>/LC_MESSAGES/foo.po file need to be
```
Comment (chrisbra) [inline runtime/doc/repeat.txt:769]: do we need to specify in what <lang_id> should look like?

File: runtime/doc/repeat.txt [hunk 7]
```diff
@@ -762,6 +765,35 @@ the command after changing the plugin help: >
...
+	:echo gettext("Hello", "foo")
+>
+to get the text "Hello" translated to the user's preferred language (if the
+plugin messages have been translated to this language).
+
+If you need to use the message in a different encoding from the current one,
+you can use (for instance) >
+	:echo gettext("Hello", "foo", "utf-8")
+<or >
+	:echo gettext("Hello", "foo", "cp1251")
```
Comment (chrisbra) [inline runtime/doc/repeat.txt]: so the returned encoding is in `cp1251`?

File: runtime/doc/repeat.txt [hunk 8]
```diff
@@ -762,6 +765,35 @@ the command after changing the plugin help: >
...
+
+To create the foo.po file, you need to create a foo.pot file first. The
+entries in this file need to be translated to the language(s) you want to be
+supported by your plugin.
+
+To create the foo.pot file, run the following command: >
+	cd ~/.vim/pack/start/foobar
+	make -f ~/src/vim/src/po/Makefile PACKAGE=foo PO_BASEDIR=~/src/vim/src/po PO_INPUTLIST= PO_VIM_JSLIST="plugin__foo.js plugin__bar.js autoload__foo.js" PO_VIM_INPUTLIST="plugin/foo.vim plugin/bar.vim autoload/foo.vim" foo.pot
+<
```
Comment (chrisbra) [inline runtime/doc/repeat.txt:792]: that sounds difficult for a user, that just wants to translate some messages.

File: src/po/Make_all.mak [hunk 9]
```diff
@@ -189,8 +189,8 @@ PO_VIM_INPUTLIST = \
 	../../runtime/defaults.vim
 
 PO_VIM_JSLIST = \
-	optwin.js \
-	defaults.js
+	________runtime__optwin.js \
+	________runtime__defaults.js
```
Comment (chrisbra) [inline src/po/Make_all.mak:193]: why are those files renamed?

**Comment by chrisbra** [comment]
Just wondering, do we really need the encoding option? Wouldn't it make more sense to always use whatever Vims default encoding is? Or otherways around it, is Vim really capable to handle vimscript strings in a different encoding other than the default?

File: Filelist (hunk 5 above)
Comment (cvwillegen) [reply Filelist:224]: Maybe, but I don't know where it's gone :-/

File: runtime/doc/repeat.txt (hunk 6 above)
Comment (cvwillegen) [reply runtime/doc/repeat.txt:769]: In "mlang.txt" there is this nugget:
```Message files (vim.mo) have to be placed in "$VIMRUNTIME/lang/xx/LC_MESSAGES",
where "xx" is the abbreviation of the language (mostly two letters).
```

Perhaps I can link to *multilang-messages* here??

File: runtime/doc/repeat.txt (hunk 7 above)
Comment (cvwillegen) [reply runtime/doc/repeat.txt]: Yes, it is.

File: src/po/Make_all.mak (hunk 9 above)
Comment (cvwillegen) [reply src/po/Make_all.mak:193]: I needed the directories that those files reside in, so that both a plugin/foo.js _and_ an autoload/foo.js can be read for messages to translate. Besides, these are temporary files...

File: runtime/doc/repeat.txt (hunk 8 above)
Comment (cvwillegen) [reply runtime/doc/repeat.txt:792]: It's not the translation per se, it's the plugin writer that wants to have their messages translated to another language... and for that, the Vim script needs to be 'converted' to "Javascript" in order to be read by the xgettext program. And I wasn't really happy with the way to run the Makefile, but I couldn't think of another way to get the plumbing to work...

**Comment by cvwillegen** [comment]
As far as I can see, I've fixed all remarks. The error on MacOS looks unrelated.

**Comment by chrisbra** [comment]
ah sorry, I always wanted to come back here and then it slipped through. I'll go through this very soon. Good we updated it, so it is moved back to the top of the *latest activity view* :)

File: runtime/doc/repeat.txt [hunk 10]
```diff
@@ -762,6 +765,29 @@ the command after changing the plugin help: >
...
 	:helptags path/opt/fooextra/doc
 
+The messages that are in the lang/<lang_id>/LC_MESSAGES/foo.po file need to be
+translated to a format that the |gettext()| function understands by running the
+*msgfmt* program. This will result in a lang/<lang_id>/LC_MESSAGES/foo.mo
+file. See |multilang| on how to specify languages.
+
+In your plugin, you need to call the |bindtextdomain()| function as follows.
+This assumes that the directory structure is as above: >
+	:call bindtextdomain("foo", getcwd()~"../lang")
```
Comment (chrisbra) [inline runtime/doc/repeat.txt]: that `getcwd()~"../lang"` looks wrong. I guess what is meant is more like this (assuming it should point to the lang directory in a directory just above where the current script is located)?
```
fnamemodify(expand("<script>"), ':p:h') .. '/../lang/')
```

File: runtime/doc/builtin.txt [hunk 11]
```diff
@@ -1169,6 +1172,13 @@ balloon_split({msg})					*balloon_split()*
 <		{only available when compiled with the |+balloon_eval_term|
 		feature}
 
+bindtextdomain({package} [, {path}])			*bindtextdomain()*
+		Bind a specific {package} to a {path} so that the
+		|gettext()| function can be used to get language-specific
+		translations for a package.
```
Comment (chrisbra) [inline runtime/doc/builtin.txt]:
```suggestion
		translations for a package.  {path} should typically use the lang directory see |package-create|
```

File: runtime/doc/builtin.txt [hunk 12]
```diff
@@ -1169,6 +1172,13 @@ balloon_split({msg})					*balloon_split()*
...
+
+		When {path} is omitted, the Vim runtime path is used.
```
Comment (chrisbra) [inline runtime/doc/builtin.txt]:
```suggestion
		When {path} is omitted, then $VIMRUNTIME/path is used.
```

File: src/evalfunc.c [hunk 13]
```diff
@@ -5825,11 +5865,43 @@ f_gettagstack(typval_T *argvars, typval_T *rettv)
...
+	rettv->vval.v_string = vim_strsave((char_u *)dgettext((const char *)argvars[1].vval.v_string, (const char *)argvars[0].vval.v_string));
+#else
+	textdomain((const char *)argvars[1].vval.v_string);
+	rettv->vval.v_string = vim_strsave((char_u *)_(argvars[0].vval.v_string));
+	textdomain(VIMPACKAGE);
+#endif
+
+#if defined(HAVE_BIND_TEXTDOMAIN_CODESET)
+	if (prev != NULL)
+	{
```
Comment (chrisbra) [inline src/evalfunc.c]: style, you can remove braces around single statements.

File: src/evalfunc.c [hunk 14]
```diff
@@ -5825,11 +5865,43 @@ f_gettagstack(typval_T *argvars, typval_T *rettv)
...
+	    bind_textdomain_codeset((const char *)argvars[1].vval.v_string, prev);
+	}
+#endif
+    }
+    else
+    {
+	rettv->vval.v_string = vim_strsave((char_u *)_(argvars[0].vval.v_string));
+    }
```
Comment (chrisbra) [inline src/evalfunc.c]: style, remove braces

File: src/evalfunc.c [hunk 15]
```diff
@@ -6230,6 +6302,13 @@ f_has(typval_T *argvars, typval_T *rettv)
 		1
 #else
 		0
+#endif
+		},
+	{"bind_codeset",
```
Comment (chrisbra) [inline src/evalfunc.c]: This needs to be documented at `:h feature-list` in builtin.txt

File: src/config.h.in [hunk 16]
```diff
@@ -424,6 +423,12 @@
 /* Define if there is a working gettext(). */
 #undef HAVE_GETTEXT
 
+/* Define if there is a working bind_textdomain_codeset(). */
+#undef HAVE_BIND_TEXTDOMAIN_CODESET
```
Comment (chrisbra) [inline src/config.h.in:427]: Is there a missing configure check for this `HAVE_BIND_TEXTDOMAIN_CODESET`?

File: src/testdir/test_gettext_cp1251.vim [hunk 17]
```diff
@@ -0,0 +1,13 @@
+source check.vim
+
+" Test for gettext()
+func Test_gettext()
+  if has('bind_codeset')
+    set encoding=cp1251
```
Comment (chrisbra) [inline src/testdir/test_gettext_cp1251.vim]: please don't change the encoding, perhaps you need `:scriptencoding` instead?

**Comment by chrisbra** [comment]
Oh, and can you please also update the vim script function descriptions and add the expected return type please?

File: runtime/doc/repeat.txt (hunk 10 above)
Comment (cvwillegen) [reply runtime/doc/repeat.txt]: Added that, thanks for the pointer! Vim script is not my forte, I'm better at the C-side of things :-/

File: runtime/doc/builtin.txt (hunk 11 above)
Comment (cvwillegen) [reply runtime/doc/builtin.txt]: Added that. Also made the {path} parameter obligatory. It's nonsense to fall back to the $VIMRUNTIME path. If you want to get a string from there, just use `gettext()` without the package...

File: runtime/doc/builtin.txt (hunk 12 above)
Comment (cvwillegen) [reply runtime/doc/builtin.txt]: {path} can no longer be omitted.

File: src/evalfunc.c (hunk 13 above)
Comment (cvwillegen) [reply src/evalfunc.c]: Ok.

File: src/evalfunc.c (hunk 14 above)
Comment (cvwillegen) [reply src/evalfunc.c]: Fixed as well.

File: src/config.h.in (hunk 16 above)
Comment (cvwillegen) [reply src/config.h.in:427]: No, it was already there, but I moved it closer to the ``HAVE_GETTEXT`` and ``HAVE_DGETTEXT`` definitions (they are used together)

File: src/testdir/test_gettext_cp1251.vim (hunk 17 above)
Comment (cvwillegen) [reply src/testdir/test_gettext_cp1251.vim]: No, I really need to change the ``encoding``, because the string is to be sent to the terminal after this. I re-set the encoding in the test file, though...

File: src/config.h.in (hunk 16 above)
Comment (chrisbra) [reply src/config.h.in:427]: I was talking about `configure.ac`. I only see a test for `dgettext()`, but not for `bind_textdomain_codeset()`?

File: src/testdir/test_gettext_cp1251.vim (hunk 17 above)
Comment (chrisbra) [reply src/testdir/test_gettext_cp1251.vim]: Hm, IIRC, setting encoding in the active vim session may invalidate all stored string data. So this is really a bad idea to change the encoding.

File: src/config.h.in (hunk 16 above)
Comment (cvwillegen) [reply src/config.h.in:427]: It's in configure.ac#L4499. I didn't change that line...

Comment (chrisbra) [reply src/config.h.in:427]: ah thanks!

File: src/testdir/test_gettext_cp1251.vim (hunk 17 above)
Comment (cvwillegen) [reply src/testdir/test_gettext_cp1251.vim]: This is 'just' a test file. In the wild, you wouldn't change ``encoding`` on the fly (I expect this to be set in the user's (or system-wide!) .vimrc). But, I need to see that my changes work, and I see no other way to test this...

Always using UTF-8 isn't going to work if the user's terminal works in cp1251 (or does it???), and both you and Bram were against specifying the encoding specifically, so changing the encoding is the only way to test this change.

I'm happy to be proven wrong, though, since this _is_ an 'ugly' bit of code, and hard to test without having a terminal (or language!) using these encodings.

**Comment by RestorerZ** [comment]
Yeah, basically, it works.
But there are a few things
1. Christian, it bad that you removed support for the encoding parameter. It is possible that when you switch the encoding in running Vim, the translated text will not be displayed correctly. It should be checked.
2. Conversion from plugin to js-file and POT file in Make_mvc.mak does not work (Macros as in Makefile have been added). I'll see what else I can do.
3. very complicated make command to create a POT file. But there is probably no other way to do it.
And a couple or three other nuances, but in general, as I said, it works wonderfully.  

Unfortunately, other things are piling up right now (so I apologize for not responding right away), but I'll get them done and continue working on Vim.

Ubuntu22x64-2024-06-19-23-42-20

**Comment by RestorerZ** [comment]
It might be better to make a separate target for the make-files to prepare a POT file for plugins.
I'll have to make a note of what to do.

**Comment by chrisbra** [comment]
> * Christian, it bad that you removed support for the encoding parameter. It is possible that when you switch the encoding in running Vim, the translated text will not be displayed correctly. It should be checked.
> 
> * Conversion from plugin to js-file and POT file in Make_mvc.mak does not work (Macros as in Makefile have been added). I'll see what else I can do.
> 
> * very complicated make command to create a POT file. But there is probably no other way to do it.
>   And a couple or three other nuances, but in general, as I said, it works wonderfully.

Well, the problem with 1) is, that I think Vim cannot handle character data with different encodings. I have no idea how this may break in all kind of funny ways. Let's not try to find this out. Also, you should never change the encoding of your running Vim instance, this may make all your existing character data invalid.

For 3) I am fully aware, but at least for now it seems not easily possible to simplify. 

Okay then let me merge this and we can make further improvements in a followup PR then.

**Comment by chrisbra** [comment]
@cvwillegen I just merged it, but something must gone wrong. The test seems to pass for me locally, but in CI they seem to fail. Even when I run them interactively, it `gettext('ERROR:', '__PACKAGE__')` just returns `ERROR` for me

**Comment by chrisbra** [comment]
I am disabling the test for now on MS-Windows and MacOS

**Comment by cvwillegen** [comment]
Op vr 21 jun. 2024 07:33 schreef Christian Brabandt <
***@***.***>:

> I am disabling the test for now on MS-Windows and MacOS
>

I pulled out my (spare) laptop, but that has OS 10.13.6, and there both
(all) gettext() tests succeed... not sure how to proceed.

Christ van Willegen

>

**Comment by LemonBoy** [comment]
> I am disabling the test for now on MS-Windows and MacOS

On Windows you need to supply the `libintl.dll` (or `libintl-8.dll`) library somehow, the test runs fine once you do so.
You can download a copy from [here](http://mlocati.github.io/articles/gettext-iconv-windows.html).

**Comment by cvwillegen** [comment]
Op vr 21 jun. 2024 16:11 schreef Christian Brabandt <
***@***.***>:

> I also noticed that the return value of gettext() and bindtext() may not
> be optimal. Perhaps it should return a number indicating success/failure?
>

dgettext() returns either the original (untranslated) msgid, or its
translation. Also, returning either True or a String would be a nightmare...

bindtextdomain() returns either NULL or the previous path that was set. We
could return True/False for success/ failure of memory allocation within
the function...

Christ van Willegen

>

**Comment by chrisbra** [comment]
Yeah, we cannot change `gettext()` function. But for `bindtextdomain()` returning true false should be useful I think, at least better than nothing :)
